src/v/datalake/translation_task.cc

```diff
@@ -0,0 +1,182 @@
+translation_task::translate(
+
+  std::unique_ptr<data_writer_factory> writer_factory,
+  model::record_batch_reader reader,
+  const remote_path& remote_path_prefix,
+  retry_chain_node& rcn,
+  lazy_abort_source lazy_as) {
+    record_multiplexer mux(std::move(writer_factory));
+    // Write local files
+    auto mux_result = co_await reader.consume(
+      std::move(mux), _read_timeout + model::timeout_clock::now());
+
+    if (mux_result.has_error()) {
+        vlog(
+          datalake_log.warn,
+          "Error writing data files - {}",
+          mux_result.error());
+        co_return errc::file_io_error;
+    }
+    auto write_result = std::move(mux_result.value());
+    vlog(
+      datalake_log.trace,
+      "translation result base offset: {}, last offset: {}, data files: {}",
+      write_result.start_offset,
+      write_result.last_offset,
+      fmt::join(write_result.data_files, ", "));
+
+    coordinator::translated_offset_range ret{
+      .start_offset = mux_result.value().start_offset,
+      .last_offset = mux_result.value().last_offset,
+    };
+    ret.files.reserve(write_result.data_files.size());
+    std::optional<errc> upload_error;
+    // TODO: parallelize uploads
+    for (auto& lf_meta : write_result.data_files) {
+        auto r = co_await execute_single_upload(
+          lf_meta, remote_path_prefix, rcn, lazy_as);
+        if (r.has_error()) {
+            vlog(
+              datalake_log.warn,
+              "error uploading file {} to object store - {}",
+              lf_meta,
+              r.error());
+            upload_error = r.error();
+            /**
+             * For now we value simplicity, therefore in case of cloud error we
+             * invalidate the whole translation i.e. we are going to cleanup all
+             * the local data files and remote files that were already
+             * successfully uploaded. Coordinator will simply retry translating
+             * the same range
+             */
+            break;
+        }
+        ret.files.push_back(coordinator::data_file{
+          .remote_path = r.value()().string(),
+          .row_count = lf_meta.row_count,
+          .file_size_bytes = lf_meta.size_bytes,
+          .hour = lf_meta.hour,
+        });
+    }
+
+    auto delete_result = co_await delete_local_data_files(
+      write_result.data_files);
+    // for now we simply ignore the local deletion failures
+    if (delete_result.has_error()) {
+        vlog(
+          datalake_log.warn,
+          "error deleting local data files - {}",
+          delete_result.error());
+    }
+
+    if (upload_error) {
+        // in this case we delete any successfully uploaded remote files before
+        // returning a result
+        chunked_vector<remote_path> files_to_delete;
+        for (auto& data_file : ret.files) {
+            files_to_delete.emplace_back(data_file.remote_path);
+        }
+        auto remote_del_result = co_await _cloud_io->delete_data_files(
```

Comment:
nit: may be add a todo for cleaning up orphan files in the bucket? eg: if deletion stops halfway  because node is shutting down or some such.